src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/MacroParserImpl.java

```diff
@@ -230,34 +127,242 @@ public ClangReparser(TranslationUnit tu, Collection<String> args) throws IOExcep
+
+            @Override
+            boolean isRecoverableFailure() {
+                return true;
+            }
+        }
+
+        class Success extends Entry {
+            final Type type;
+            final Object value;
+
+            public Success(String name, String[] tokens, Cursor cursor, Type type, Object value) {
+                super(name, tokens, cursor);
+                this.type = type;
+                this.value = value;
+            }
+
+            @Override
+            boolean isSuccess() {
+                return true;
+            }
+
+            public Object value() {
+                return value;
+            }
+        }
+
+        void enterMacro(String name, String[] tokens, Cursor cursor) {
+            Unparsed unparsed = new Unparsed(name, tokens, cursor);
+            macrosByMangledName.put(unparsed.mangledName(), unparsed);
+        }
+
+        public List<Declaration.Constant> reparseConstants() {
+            int last = -1;
+            while (macrosByMangledName.size() > 0 && last != macrosByMangledName.size()) {
+                last = macrosByMangledName.size();
+                // step 1 - try parsing macros as var declarations
+                reparseMacros(false);
+                // step 2 - retry failed parsed macros as pointers
+                reparseMacros(true);
+            }
+            treeMaker.typeMaker.resolveTypeReferences();
+            return macrosByMangledName.values().stream()
+                    .filter(Entry::isSuccess)
+                    .map(e -> treeMaker.createMacro(e.cursor, e.name, ((Success)e).type, ((Success)e).value))
+                    .collect(Collectors.toList());
+        }
+
+        void updateTable(TypeMaker typeMaker, Cursor decl) {
+            String mangledName = decl.spelling();
+            Entry entry = macrosByMangledName.get(mangledName);
+            Entry newEntry;
+            try (EvalResult result = decl.eval()) {
+                switch (result.getKind()) {
+                    case Integral: {
+                        long value = result.getAsInt();
+                        newEntry = entry.success(typeMaker.makeType(decl.type()), value);
+                        break;
+                    }
+                    case FloatingPoint: {
+                        double value = result.getAsFloat();
+                        newEntry = entry.success(typeMaker.makeType(decl.type()), value);
+                        break;
+                    }
+                    case StrLiteral: {
+                        String value = result.getAsString();
+                        newEntry = entry.success(typeMaker.makeType(decl.type()), value);
+                        break;
+                    }
+                    default: {
+                        Type type = decl.type().equals(decl.type().canonicalType()) ?
+                                null : typeMaker.makeType(decl.type());
+                        newEntry = entry.failure(type);
```

Comment:
Not sure what's going on here exactly, but AFAICS passing null to `failure` will either return null, making newEntry null, or throw an exception.

Would it be clearer to assign null to newEntry directly? (Also, maybe using a switch expression?)

Reply:
Not sure I understand. If the old entry is an UnparsedMacro, then failure(null) is != from failure(T) where T != null; the former returns null (meaning no entry), while the latter creates a new recoverable failure entry. The exception you mention is thrown regardless of the type - and occurs when failure is called on an entry for which no state transition associated to failure exists.

I can update the code to use a switch expression, but I think that encapsulating the state transition logic into the methods of the entry leads to better code - this is essentially an OO implementation of a finite state machine.